File: src/cmcandy/Java_Language_Answers/_0022.java
package cmcandy.Java_Language_Answers;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by CHEN on 2020/4/21.
 */
public class _0022 {

    static class Solution {
        List<String> res = new ArrayList<>();
        public List<String> generateParenthesis(int n) {
            dfs(n, n, "");
            return res;

        }

        private void dfs(int left, int right, String s) {

            if (left==0&&right==0){
                res.add(s);
            }
            else {
                if (left>0){
                    dfs(left-1,right,s+"(");
                }if (right>left){
                    dfs(left,right-1,s+")");
                }
            }
        }
    }
}

File: src/cmcandy/Python_language_Answers/_0022.py
from typing import List


class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        self.res = []
        self.generator('', n, n)
        return self.res
        pass

    def generator(self, s: str, left: int, right: int):
        '''
            left代表当前有多少个(
            right代表当前有多少个)
        '''
        if left == 0 and right == 0:
            self.res.append(s)
        elif left < right:
            self.generator(s + ')', left, right - 1)
            if left >= 1:
                self.generator(s + '(', left - 1, right)
        else:
            self.generator(s + '(', left - 1, right)


if __name__ == '__main__':
    s = Solution()
    print(s.generateParenthesis(3))